internal/types/compare.go

```diff
@@ -255,3 +251,121 @@ func compareNumbers(a float64, b int64) CompareResult {
 
 	return CompareResult(bigA.Cmp(bigB))
 }
+
+// dataTypeOrderResult represents the comparison order of data types.
+type dataTypeOrderResult uint8
+
+const (
+	_ dataTypeOrderResult = iota
+	minKeyDataType
+	nullDataType
+	nanDataType
+	numbersDataType
+	stringDataType
+	objectDataType
+	arrayDataType
+	binDataType
+	objectIdDataType
+	booleanDataType
+	dateDataType
+	timestampDataType
+	regexDataType
+	maxKeyDataType
+)
+
+// defineDataType define which type has value and returns a sequence the type has.
+func defineDataType(value any) dataTypeOrderResult {
+	switch value := value.(type) {
+	case NullType:
+		return nullDataType
+	case float64:
+		if math.IsNaN(value) {
+			return nanDataType
+		}
+		return numbersDataType
+	case int32, int64:
+		return numbersDataType
+	case string:
+		return stringDataType
+	case bool:
+		return booleanDataType
+	case Binary:
+		return binDataType
+	case ObjectID:
+		return objectIdDataType
+	case time.Time:
+		return dateDataType
+	case Timestamp:
+		return timestampDataType
+	case Regex:
+		return regexDataType
+	}
+
+	panic(fmt.Sprintf("value cannot be defined, value is %v, data type of value is %T", value, value))
+}
+
+// numberDataTypeOrderResult represents the comparison order of numbers.
+type numberDataTypeOrderResult uint8
+
+const (
+	_ numberDataTypeOrderResult = iota
+	doubleNegativeZero
+	doubleDT
+	int32DT
+	int64DT
+)
+
+// defineNumberDataType define which number type has value and returns a sequence the type has.
+func defineNumberDataType(value any) numberDataTypeOrderResult {
+	switch value := value.(type) {
+	case int64:
```

Comment:
Please rearrange cases in order with [code style](https://github.com/FerretDB/FerretDB/blob/main/CONTRIBUTING.md#code-style-and-conventions)

Reply:
good catch! thank you! the cases have been rearrangeg.